src/main/java/org/mule/service/http/impl/service/client/NonBlockingStreamWriter.java

```diff
@@ -0,0 +1,190 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+
+/**
+ * Writes the data passed via the non-blocking method {@link #addDataToWrite(OutputStream, byte[], Supplier)} to the specified
+ * stream, only if there is available space. If no space available, it saves a task to try later. When it tried to execute all the
+ * pending writes, and it couldn't, then it sleeps a certain period passed via constructor (default is
+ * {@link #DEFAULT_TIME_TO_SLEEP_WHEN_COULD_NOT_WRITE_MILLIS}). This is done to avoid an unnecessarily high CPU consumption.
+ */
+public class NonBlockingStreamWriter implements Runnable {
+
+  private static final Logger LOGGER = getLogger(NonBlockingStreamWriter.class);
+  private static final int DEFAULT_TIME_TO_SLEEP_WHEN_COULD_NOT_WRITE_MILLIS = 100;
+
+  private final AtomicBoolean isStopped = new AtomicBoolean(false);
+  private final BlockingQueue<InternalWriteTask> tasks = new LinkedBlockingQueue<>();
+  private final int timeToSleepWhenCouldNotWriteMillis;
+
+  public NonBlockingStreamWriter(int timeToSleepWhenCouldNotWriteMillis) {
+    this.timeToSleepWhenCouldNotWriteMillis = timeToSleepWhenCouldNotWriteMillis;
+  }
+
+  public NonBlockingStreamWriter() {
+    this(DEFAULT_TIME_TO_SLEEP_WHEN_COULD_NOT_WRITE_MILLIS);
+  }
+
+  /**
+   * Tries to write from <code>dataToWrite</code> to <code>destinationStream</code>, as many bytes as the
+   * <code>availableSpace</code> supplier says it's possible to write. If the supplier returns <code>0</code>, it schedules a task
+   * to try later.
+   * 
+   * @param destinationStream where the data has to be written.
+   * @param dataToWrite       the data to write.
+   * @param availableSpace    a supplier that says how many bytes can be written to the stream without blocking.
+   * @return a {@link CompletableFuture} that will be completed when all the data was written, or when an exception occurs.
+   */
+  public CompletableFuture<Void> addDataToWrite(OutputStream destinationStream,
+                                                byte[] dataToWrite,
+                                                Supplier<Integer> availableSpace) {
+
+    InternalWriteTask internalWriteTask = new InternalWriteTask(destinationStream, dataToWrite, availableSpace);
+    boolean couldCompleteSync = internalWriteTask.execute();
+    if (!couldCompleteSync) {
+      tasks.add(internalWriteTask);
+    }
+    return internalWriteTask.getFuture();
+  }
+
+  @Override
+  public void run() {
+    try {
+      while (!isStopped.get()) {
+        boolean couldWriteSomething = writeWhateverPossible();
+
+        if (!couldWriteSomething && !isStopped.get()) {
+          LOGGER.trace("Giving some time to the other threads to consume from pipes...");
+          sleep(timeToSleepWhenCouldNotWriteMillis);
+        }
+      }
+    } catch (InterruptedException e) {
+      currentThread().interrupt();
+    }
```

Comment:
If the thread is interrupted while sleeping or polling, we will be exiting the loop, finishing the thread, even if `isStopped` is still `false`. Are we sure that is what we want? or should we re-check the `isStopped` flag?

From what I saw in the tests, this seems to be intentional. But couldn't it cause issues if it wasn't just a manual interruption of the thread?